internal/gatewayapi/helpers.go

```diff
@@ -495,3 +496,37 @@ func irConfigName(policy client.Object) string {
 		strings.ToLower(policy.GetObjectKind().GroupVersionKind().Kind),
 		utils.NamespacedName(policy).String())
 }
+
+func getPolicyTargetRefs[T client.Object](policy egv1a1.PolicyTargetReferences, potentialTargets []T) []gwapiv1a2.LocalPolicyTargetReferenceWithSectionName {
+	dedup := map[gwapiv1a2.LocalPolicyTargetReferenceWithSectionName]byte{}
+	for _, ref := range policy.GetTargetRefs() {
+		dedup[ref] = byte(0)
+	}
+
+	for _, currSelector := range policy.TargetSelectors {
+		labelSelector := labels.SelectorFromSet(currSelector.MatchLabels)
+		for _, obj := range potentialTargets {
+			gvk := obj.GetObjectKind().GroupVersionKind()
+			if gvk.Kind != string(currSelector.Kind) ||
+				gvk.Group != string(ptr.Deref(currSelector.Group, "gateway.networking.k8s.io")) {
+				continue
+			}
+
+			if labelSelector.Matches(labels.Set(obj.GetLabels())) {
+				dedup[gwapiv1a2.LocalPolicyTargetReferenceWithSectionName{
+					LocalPolicyTargetReference: gwapiv1a2.LocalPolicyTargetReference{
+						Group: gwapiv1a2.Group(gvk.Group),
+						Kind:  gwapiv1a2.Kind(gvk.Kind),
+						Name:  gwapiv1a2.ObjectName(obj.GetName()),
+					},
+				}] = byte(0)
+			}
+		}
+	}
+
+	ret := []gwapiv1a2.LocalPolicyTargetReferenceWithSectionName{}
+	for key := range dedup {
```

Comment:
can we garantee stability of list order with maps or sets ?

Reply:
also is there a way to make this stateless ? we are actively trying to reduce mem usage https://github.com/envoyproxy/gateway/issues/3698

Reply:
> can we garantee stability of list order with maps or sets ?

No. Sets are the same as the builtin maps, since they are implemented in the same way, and they do not guarantee the list order. When transforming the map to a list, I can sort the list. What would be the correct key in which the list should be ordered?

> is there a way to make this stateless ? we are actively trying to reduce mem usage

Using a `map` to deduplicate the set of relevant targets means that the code never allocates more memory than it would need to keep track of the targets of the policy. I don't see a way to collect the targets of the selectors and not keep _some_ state. Note that the memory is only allocated for the duration of the time it takes to handle the policy - it's not some cache that is kept and never released.

Reply:
lets sort based on creation timestamp
```
	// Sort based on timestamp
	sort.Slice(ret, func(i, j int) bool {
		return ret[i].CreationTimestamp.Before(&(ret[j].CreationTimestamp))
	})
	```